### PR TITLE
Rebalance Event channel

### PR DESCRIPTION
**Pull request by wannes-ds**
This PR adds a channel to the Consumer API though which rebalance events are published, informing the user of partitions that have been assigned or revoked. It's functionally similar to the ConsumerRebalanceListener in the Java API.

This PR implements #18

### CODE REVIEW

File: consumer.go [hunk 1]
```diff
@@ -7,6 +7,18 @@ import (
 	"github.com/Shopify/sarama"
 )
 
+type RebalanceEventType int
+
+const (
+	PartitionAssigned RebalanceEventType = iota
```
Comment (dim) [inline consumer.go]: Sorry, that's not how iota works, please fix, it should be:
```
const (
	PartitionAssigned RebalanceEventType = iota
	PartitionRevoked
)
```

File: consumer.go [hunk 2]
```diff
@@ -7,6 +7,18 @@ import (
 	"github.com/Shopify/sarama"
 )
 
+type RebalanceEventType int
```
Comment (dim) [inline consumer.go]: Not sure, does it really need to be an `int`, why not a `uint8`?

File: consumer.go [hunk 3]
```diff
@@ -264,6 +285,10 @@ func (c *Consumer) rebalance() error {
 		return err
 	}
 
+	adds, rems := Diff(oldSubs, subs)
```
Comment (dim) [inline consumer.go]: I think we should a configuration option for that, something like `Config.Group.Return.RebalanceEvents`, defaulting to false, in line with sarama standard configuration

File: partitions.go [hunk 4]
```diff
@@ -203,3 +203,60 @@ func (m *partitionMap) Info() map[string][]int32 {
 	}
 	return info
 }
+
+// Diffs map with current partitionMap, returns additions and removals
+func Diff(current map[string][]int32, new map[string][]int32) (map[string][]int32, map[string][]int32) {
```
Comment (dim) [inline partitions.go]: Does `Diff` really need to be a public function? Please protect, a unit test would be good too

File: consumer.go (hunk 1 above)
Comment (dim) [inline consumer.go]: Also, do we need them at all? Can't we have just one:

```go
type RebalanceEvent struct {
	Assigned, Revoked map[string][]int32
}
```

**Comment by wannes-ds** [comment]
After checking it seems the Java API revokes all partition assignments *before* a rejoin is attempted. Adjusted this PR to reflect this, removing the Diff function.

`RebalanceEventType` is fixed, because revocations and assignments are now called at different points in time, I believe this constant has its use, though I'll happily change it to something more to your liking.

Would you still prefer to have the config option, as no additional processing needs to take place anymore to provide these events.

**Comment by dim** [comment]
@wannes-ds how useful is it to be in sync with the Java API? I would actually prefer to emit an object with 3 maps to the channel, something like:

```go
type RebalanceEvent struct {
  Claimed, Revoked, Current map[string][]int32
}
```
Seems much more useful and informative. 

So yes, I would like to have a config option. Also, `rebal` needs to be closed on exit.

**Comment by dim** [comment]
Superseded by #21

**Pull request closed without merging**